Fix #5872: Buf Lint Script Does Not Resolve Darwin-arm64 Execs for Apple Silicon Macs

harleenkaur2003 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #5872: Buf Lint Script Does Not Resolve Darwin-arm64 Execs for Apple Silicon Macs


This PR fixes the issue where the `buf_lint_check.sh` script didn't handle the `Darwin-arm64` architecture for Apple Silicon Macs. The script has been updated to detect the architecture and download the appropriate `buf` binary. Additionally, the `BUFVERSION` has been updated to `v0.42.0` to support the `Darwin-arm64` architecture.

**Changes made:**
- Updated the script to handle `Darwin-arm64` architecture.
- Updated the `BUFVERSION` to `v0.42.0` to support Apple Silicon Macs.
- Fixed issues with missing binary for Apple Silicon Macs.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).


If there are any further suggestions regarding my changes or if any adjustments are needed to align with the style guide or script functionality, kindly let me know.

harleenkaur2003 committed:
Fix buf lint script to resolve Darwin-arm64 execs for Apple Silicon Macs

harleenkaur2003 committed:
Update BUFVERSION to v0.42.0 to support Darwin-arm64 architecture

harleenkaur2003 commented:
Hi maintainers,
This is my first contribution to the repo. I’ve implemented the fix as discussed in the issue:

Updated buf_lint_check.sh to handle the Darwin-arm64 architecture.

Updated BUFVERSION to v0.42.0 in buf_download.sh to support Apple Silicon Macs.

I haven’t run the pre-push checks or tested the change on a Mac device since I’m currently on Windows.
I’ve tried to follow the existing formatting and style in the surrounding code.

Please let me know if any specific changes or tests are required. I’ll be happy to update the PR accordingly. Thanks!

adhiamboperes commented:
PTAL @manas-yu for a first review pass.

manas-yu reviewed (commented):
Thanks @harleenkaur2003, just a minor comment, also could you please update the PR description and title to include the issue with "Fix #bugnum: "

manas-yu commented on scripts/buf_lint_check.sh:
```diff
@@ -32,6 +32,8 @@ populate_jar_config_file_paths() {
 
   if [ "$1" = "Linux" ]; then
     buf_file_name="buf-Linux-x86_64"
+  elif [ "$1" = "Darwin" ] && [ "$(uname -m)" = "arm64" ]; then
+    buf_file_name="buf-Darwin-arm64"
   else
     buf_file_name="buf-Darwin-x86_64"
   fi
```
```suggestion
case "$1" in
  Linux)
    buf_file_name="buf-Linux-x86_64"
    ;;
  Darwin)
    if [ "$(uname -m)" = "arm64" ]; then
      buf_file_name="buf-Darwin-arm64"
    else
      buf_file_name="buf-Darwin-x86_64"
    fi
    ;;
  *)
    echo "Protobuf lint check not available on $1"
    exit 0
    ;;
esac
```
 Nit: we could use `case` block instead of `if/elif` block, would be more cleaner.

harleenkaur2003 committed:
Updated buf lint script to support Darwin-arm64 architecture

harleenkaur2003 commented:
Hi @manas-yu ,

I've noticed that my PR now shows a large number of unrelated file changes and merge conflicts, possibly due to recent updates in the develop branch after I opened the PR. My changes were originally limited to just the buf lint script.

Should I try resolving all the merge conflicts manually, or should I create a fresh PR with only the intended changes based on the latest develop?

Please let me know what you'd suggest.

manas-yu commented:
Hi @harleenkaur2003, I noticed that your [branch](https://github.com/harleenkaur2003/oppia-android/tree/fix/build-setup-compatibility) isn't up-to-date. It is essential to keep your branch in sync with the develop branch. You may need to fetch changes from `upstream/develop` to your branch, then push your changes again. Let me know if you run into issues afterwards.

Also, it's just something we follow here, don't resolve comments instead, reply with an acknowledgement like done or updated, and also include the Bugnum in the PR description as well(eg: Fixes #Bugnum). Thanks!

harleenkaur2003 commented:
Hi @manas-yu ,

I've synced my branch with the latest changes from develop and force-pushed the updated branch. I’ve also updated the PR title and description to include the correct Fixes #5872 reference.

Additionally, I’ve implemented the changes you suggested regarding using a case block instead of if/elif.
Please take a look when you get a chance. Let me know if any further updates are needed; I’d be happy to address them.
Thanks again!

manas-yu commented:
Thanks @harleenkaur2003, changes look good to me, although try avoiding force pushing changes and keeping your fork branch updated regularly :)

@adhiamboperes PTAL, also if you could test it on mac it would be great since I'm on linux. Thanks!

harleenkaur2003 commented:
Thank you @manas-yu! I really appreciate your guidance and reviews. I’ll be sure to keep these points in mind moving forward.

adhiamboperes reviewed (changes requested):
Thanks @harleenkaur2003 for the excellent PR! and thanks @manas-yu for the review pass.

@harleenkaur2003, I have tested the changes locally eand everything looks good. However, I cannot merge the PR because the branch was force pushed. Force pushing rewrites version history, therefore we do not allow it.

Here are your next steps:
1. On your local terminal, check out  a new branch from this one, e.g,   `git checkout -b fix-build-setup-compatibility`
3. Push the changes upstream
4. Create a new PR, copy the PR title and description from this current one
5. Assign me directly
6. Close this original PR

Your new PR should be merged smoothly.  

Thanks!

harleenkaur2003 commented:
Thank you @adhiamboperes for your guidance! I’ll follow the steps you outlined and create a new PR accordingly.

harleenkaur2003 commented:
#5889. Closing this one.
